sdk/tests/conformance/buffers/buffer-data-and-buffer-sub-data.html

```diff
@@ -0,0 +1,163 @@
+
+var wtu = WebGLTestUtils;
+var gl = wtu.create3DContext();
+var buf = null;
+var array = null;
+var arrayZero = null;
+if (!gl) {
+    testFailed("WebGL context does not exist");
+} else {
+    testPassed("WebGL context exists");
+
+    bufferDataTest();
+    bufferDataTestWithArrayBuffer();
+
+    bufferSubDataTest();
+}
+
+function bufferDataTest() {
+    debug("");
+    debug("Test bufferData without ArrayBuffer input");
+
+    buf = gl.createBuffer();
+    shouldBeNonNull("buf");
+
+    gl.bufferData(gl.ARRAY_BUFFER, 4, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, buf);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferData(gl.ARRAY_BUFFER, -4, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+
+    gl.bufferData(gl.ARRAY_BUFFER, 0, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferData(gl.ARRAY_BUFFER, 4, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, null);
+}
+
+function bufferDataTestWithArrayBuffer() {
+    debug("");
+    debug("Test bufferData with ArrayBuffer input");
+
+    array = new ArrayBuffer(128);
+    shouldBeNonNull("array");
+
+    buf = gl.createBuffer();
+    shouldBeNonNull("buf");
+
+    gl.bufferData(gl.ARRAY_BUFFER, array, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, buf);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferData(gl.ARRAY_BUFFER, null, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+
+    // This should not crash, but the selection of the overload is ambiguous per Web IDL.
+    gl.bufferData(gl.ARRAY_BUFFER, null, gl.STATIC_DRAW);
+    gl.getError();
+
+    gl.bufferData(gl.ARRAY_BUFFER, array, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    arrayZero = new ArrayBuffer(0);
+    shouldBeNonNull("arrayZero");
+    gl.bufferData(gl.ARRAY_BUFFER, arrayZero, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, null);
+}
+
+function bufferSubDataTest() {
+    debug("");
+    debug("Test bufferSubData");
+
+    buf = gl.createBuffer();
+    gl.bindBuffer(gl.ARRAY_BUFFER, buf);
+    gl.bufferData(gl.ARRAY_BUFFER, 128, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    array = new ArrayBuffer(64);
+
+    gl.bufferSubData(gl.ARRAY_BUFFER, -10, array);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+
+    gl.bufferSubData(gl.ARRAY_BUFFER, 65, array);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE, "buffer overflow");
+
+    gl.bufferSubData(gl.ARRAY_BUFFER, -10, new Float32Array(8));
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+
+    gl.bufferSubData(gl.ARRAY_BUFFER, 10, array);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferSubData(gl.ARRAY_BUFFER, 10, null);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
```

Comment:
This should generate INVALID_VALUE per https://www.khronos.org/registry/webgl/specs/latest/1.0/#5.14.5 . Please update the expectations, and any buggy WebGL implementations will be fixed.